components/support/sql/src/shutdown.rs

```diff
@@ -0,0 +1,81 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+License, v. 2.0. If a copy of the MPL was not distributed with this
+* file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use interrupt_support::Interruptee;
+/// Shutdown handling for database operations
+///
+/// This crate allows us to enter shutdown mode, causing all `SqlInterruptScope` instances that opt-in to
+/// shutdown support to be permanently interrupted.  This means:
+///
+///   - All current scopes will be interrupted
+///   - Any attempt to create a new scope will be interrupted
+///
+/// Here's how add shutdown support to a component:
+///
+///   - Use `SqlInterruptScope::new_with_shutdown_check()` to create a new
+///     `SqlInterruptScope`
+///   - Database connections need to be wrapped in a type that:
+///      - Implements `AsRef<SqlInterruptHandle>`.
+///      - Gets wrapped in an `Arc<>`.  This is needed so the shutdown code can get a weak reference to
+///        the instance.
+///      - Calls `shutdown::register_interrupt_handle()` on creation
+///
+///  See `PlacesDb::begin_interrupt_scope()` and `PlacesApi::new_connection()` for an example of
+///  how this works.
```

Comment:
It's kind of weird that this module is inside the `sql-support` crate, but it kind of needs to be because of how it integrates with `SqlInterruptScope` and `SqlInterruptHandler`.  I think we should consider consider moving both `shutdown.rs` and `interrupt.rs` to the `interrupt-support` crate.